framework/topology.cpp

```diff
@@ -728,24 +347,85 @@ static bool fill_ppin_sysfs(struct cpu_info *info)
     return false;
 }
 
-template <auto &fnArray> static bool try_detection(struct cpu_info *cpu)
+static bool fill_topology_hwloc(struct cpu_info *info, hwloc_topology_t topology)
+{
+    hwloc_obj_t pu = hwloc_get_pu_obj_by_os_index(topology, info->cpu_number);
+    if (!pu) {
+        hwloc_topology_destroy(topology);
+        return false;
+    }
+
+    hwloc_obj_t package = hwloc_get_ancestor_obj_by_type(topology, HWLOC_OBJ_PACKAGE, pu);
+    info->package_id = package ? package->os_index : -1;
+    hwloc_obj_t core = hwloc_get_ancestor_obj_by_type(topology, HWLOC_OBJ_CORE, pu);
+    info->core_id = core ? core->logical_index : -1;
+
+    if (core) {
+        for (unsigned i = 0; i < core->arity; ++i) {
+            if (core->children[i] == pu) {
+                info->thread_id = i;
+                break;
+            }
+        }
+    }
+
+    int depth = hwloc_get_type_depth(topology, HWLOC_OBJ_NUMANODE);
+    if (depth != HWLOC_TYPE_DEPTH_UNKNOWN) {
+        int num_nodes = hwloc_get_nbobjs_by_depth(topology, depth);
+        if (pu && pu->nodeset) {
+            for (int i = 0; i < num_nodes; i++) {
+                hwloc_obj_t numa_node = hwloc_get_obj_by_depth(topology, depth, i);
+                if (hwloc_bitmap_isset(pu->nodeset, numa_node->os_index)) {
+                    info->numa_node_id = numa_node->os_index;  // Not logical_index
+                    break;
+                }
+            }
+        }
+    }
+
+    hwloc_obj_t cache;
+    for (int level = 0; level < 3; ++level) { // Assuming 3 levels: L1, L2, L3
+        switch (level) {
+            case 0:
+                cache = hwloc_get_ancestor_obj_by_type(topology, HWLOC_OBJ_L1CACHE, pu);
+                break;
+            case 1:
+                cache = hwloc_get_ancestor_obj_by_type(topology, HWLOC_OBJ_L2CACHE, pu);
+                break;
+            case 2:
+                cache = hwloc_get_ancestor_obj_by_type(topology, HWLOC_OBJ_L3CACHE, pu);
+                break;
+        }
+        if (cache && cache->attr) {
+            if (cache->attr->cache.type == HWLOC_OBJ_CACHE_UNIFIED || cache->attr->cache.type == HWLOC_OBJ_CACHE_DATA) {
+                info->cache[level].cache_data = cache->attr->cache.size;
+            }
+            if (cache->attr->cache.type == HWLOC_OBJ_CACHE_UNIFIED || cache->attr->cache.type == HWLOC_OBJ_CACHE_INSTRUCTION) {
+                info->cache[level].cache_instruction = cache->attr->cache.size;
+            }
+        }
+    }
```

Comment:
Future thought is required here:
* cache topology is always the same inside of a given package (and usually is for the entire system)
* but we also need to know what other cores a given cache level is shared with

Nothing to do now, just leaving the comment for discussion later.